### ISSUE TITLE
suspiciously small delta argument to _process

### ISSUE THREAD

**Issue by zmanuel** [godotengine/godot#35718]
**Godot version:** 3.2

**OS/device including version:** Linux (Manjaro), NVIDIA closed source driver 430.64

**Issue description:** Prerequisites: thread_model needs to be multi-threaded, a project with low CPU load and a machine with <=4 cores probably help, physics_jitter_fix = 0, vsync enabled.
The script-visible effect is that every once in a while, _process is called with a delta far smaller than the screen refresh would suggest. The attached sample project prints for me, for example
```
0.000244
0.000197
0.000105
0.000182
0.000244
0.000202
0.000132
0.000161
0.000101
0.000137
```
after a couple of seconds.
The calls with small delta are not neighbors to calls with suspiciously large delta, as would be the case if the time of one frame was measured wrong.

**Steps to reproduce:** Run the attached sample project. To build a new project from scratch, activate multithreading, set physics_jitter_fix to 0 (positive values of that cushion the effect somewhat).

**Minimal reproduction project:** [MultithreadTimerSpikes.zip](https://github.com/godotengine/godot/files/4130551/MultithreadTimerSpikes.zip). Relevant script code just prints abnormal delta values:
```
var average_delta := 1/60.0
var warmup := 5

func _process(delta: float):
	if warmup > 0:
		warmup = warmup - 1
		return
	
	if delta < average_delta * .5 or delta > average_delta * 1.5:
		print(delta)

	average_delta = average_delta + .01 * (delta - average_delta)
```

**Comment by zmanuel** [comment]
The issue is triggered by this function:
```
void VisualServerWrapMT::thread_draw(bool p_swap_buffers, double frame_step) {

	if (!atomic_decrement(&draw_pending)) {

		visual_server->draw(p_swap_buffers, frame_step);
	}
}

```
If I put a diagnostic line into there:
```
void VisualServerWrapMT::thread_draw(bool p_swap_buffers, double frame_step) {

	if (!atomic_decrement(&draw_pending)) {

		visual_server->draw(p_swap_buffers, frame_step);
	}
	else
	{
		print_line("missed!");
	}
}
```
the 'missed!' output coincides with the script's output.

A normal game loop cycle goes like this (start point is not the Main::iteration entry point):
In sequence:
1. Main thread: Call VisualServerWrapMT::sync(), increasing draw_pending and blocking
2. Render thread: Call VisualServerWrapMT::thread_flush(), decreasing draw_pending
3. Main thread: Call VisualServerWrapMT::draw(), increasing draw_pending (not blocking)
In parallel:
4. Render thread: Call VisualServerWrapMT::thread_draw(), decreasing draw_pending, calling visual_server->draw 
5. Main thread: Do all the other stuff

If the main thread is now exceptionally non-busy in 5, it can happen that it goes around and enters step 1 again before the render thread has even begun to execute step 4. The increase of draw_pending in 1. then happens before the decrease and test of draw_pending in the previous iteration's step 4, the if condition fails and rendering the frame is dropped, leading to a reported short frame.

Seems to me like draw_pending can be eliminated completely without loss. Its purpose seems to be to prevent too many concurrent drawing related operations to happen at once/queued, but command_queue used to schedule the operations already is a FIFO with just one consumer, and VisualServerWrapMT::sync already feeds it using push_and_sync, which only returns after thread_flush is fully executed. Dunno, maybe there is an issue on some other platform I'm not considering?

Slightly less dramatic, one could remove the use of draw_pending in the flush/sync pair of methods.

I'd be happy to provide PRs for either.

Anyway, this form of the bug looks to be low priority. It really requires a project with low CPU usage where one can stick to the single threaded model without harm.
